### PR TITLE
[BAHIR-89] Multi topic support API for streaming MQTT

### PR DESCRIPTION
**Pull request by anntinutj**
New API which accept array of MQTT topics as input and return Tuple2<TopicName, Message>
as output. Help to consume from multiple MQTT with efficient user of resources.

### CODE REVIEW

**Comment by ApacheBahir** [comment]
Can one of the admins verify this patch?

File: streaming-mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala [hunk 1]
```diff
@@ -199,7 +199,181 @@ object MQTTUtils {
...
+    new MQTTPairedInputDStream(ssc, brokerUrl, topics, storageLevel, clientId, username, password,
+          cleanSession, qos, connectionTimeout, keepAliveInterval, mqttVersion)
+  }
+
+  /**
+   * Create an input stream that receives messages pushed by a MQTT publisher.
+   * Storage level of the data will be the default StorageLevel.MEMORY_AND_DISK_SER_2.
+   * @param jssc      JavaStreamingContext object
+   * @param brokerUrl Url of remote MQTT publisher
+   * @param topic     Array of topic names to subscribe to
```
Comment (ckadner) [inline streaming-mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala]: should be `@param topics` (plural)

File: streaming-mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala [hunk 2]
```diff
@@ -199,7 +199,181 @@ object MQTTUtils {
...
+    ): JavaReceiverInputDStream[(String, String)] = {
+    implicitly[ClassTag[AnyRef]].asInstanceOf[ClassTag[String]]
+    createPairedStream(jssc.ssc, brokerUrl, topics, storageLevel)
+  }
+
+  /**
+   * Create an input stream that receives messages pushed by a MQTT publisher.
+   * @param jssc               JavaStreamingContext object
+   * @param brokerUrl          Url of remote MQTT publisher
+   * @param topic              Array of topic names to subscribe to
```
Comment (ckadner) [inline streaming-mqtt/src/main/scala/org/apache/spark/streaming/mqtt/MQTTUtils.scala]: should be @param topics (plural)

File: streaming-mqtt/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+/bin/
```
Comment (ckadner) [inline streaming-mqtt/.gitignore]: why do we need this?

**Review by ckadner — CHANGES_REQUESTED** [review]
@anntinutj -- I took a quick look, still need to do a more in-depth review. Would you mind adding a test case as well (in MQTTStreamSuite.scala)?

@fbeneventi -- you added a thumbs-up, did you got a chance to verify this PR?

File: streaming-mqtt/.gitignore (hunk 3 above)
Comment (anntinutj) [reply streaming-mqtt/.gitignore]: No Need for streaming-mqtt/.gitignore , I will add Test case

**Review by anntinutj — COMMENTED** [review]
Hi @ckadner, I had done the changes, could you please review it again. Also assign bahir-89 to me and add me as a developer?

**Comment by fbeneventi** [comment]
@ckadner -- I wiil test the anntinutj PR in the next days, but I need to repalce this in my current setup. I'm just using my own modified version of the MQTT-connector that does quite the same of anntinutj but, however, it doesn't support multi-topic input and so it wastes resources allocating multiple MQTT receivers, one for each supplementary topic that has a different pattern (i.e. cannot be inserted with wildcards).

The anntinutj PR is very important because:

- preserves resources in the Spark clusters
- gives the opportunity to filter topics, an thus to efficiently manage payloads, inside the Spark programs.

**Comment by anntinutj** [comment]
Any update on the review process?

**Comment by ckadner** [comment]
@anntinutj -- Thank you for adding a test case. This looks good to me. Two thing to be added are a Python test and an example, but we could create a separate JIRA for that.

@fbeneventi -- Did you get a chance to check out this PR? we'd appreciate your comments since you seem to have a real use case for this PR.

**Comment by anntinutj** [comment]
@fbeneventi , Any updates from your side? If everything looks good @ckadner can you please merge this to master and close PR?

**Comment by fbeneventi** [comment]
I apologize for the delay, but I was away due to work.

I started to test this PR few days ago, using pyspark. After fixing some small issues, I found a problem when using the new API (MQTTUtils.createPairedStream).
I create a JIRA for this, [[BAHIR-104]](https://issues.apache.org/jira/browse/BAHIR-104?jql=project%20%3D%20BAHIR).

I don't know if the current API behavior is the expected one. I think that a small example/test with python and new multi-topic API could help a lot.

**Pull request closed without merging**